rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,392 @@
+      log.Error("while signaling: %s", err)
+    }
+
+    // If we want the database to become unavailable until 
+    // corruption is fixed, return an error and SlateDB will
+    // preform a safe shutdown.
+    return fmt.Errorf("Corruption Detected: %s", d.String())
+
+    // Else we return no error, and the database remains available.
+    return nil
+  }
+```
+
+If not defined by the user, the `DBOptions.on_corruption` callback should
+both log and return an error. This ensures that SlateDB is durable by default.
+
+#### Handling WAL Corruption
+It is reasonable for a developer to assume that once data is confirmed to be
+written to the WAL, the data or transaction is considered durably written to
+storage. However, as can be seen in the "Corruption Review" section of this
+document, this assumption is only theoretical. In practice, it is possible for
+WAL entries written to storage to be corrupted resulting in data loss.
+Additionally, this corruption may only be discovered during compaction or
+recovery.
+
+It is important to note that during normal LSM operation, the movement of
+entries from the WAL to the SSTable occurs in memory. The WAL is only used as a
+recovery mechanism in case of a catastrophic failure of the LSM. Therefore, the
+loss of committed WAL data can only occur then following two conditions are
+true.
+
+1. The LSM application exits before compaction can complete.
+2. Corruption of the WAL occurs.
+
+Although the probability of both conditions being true is very low, especially
+in the context of durable object storage, it is not impossible.
+
+To ensure durability by default, I propose that `open_with_opts()` perform
+WAL recovery synchronously and return any WAL corruption found 
+as errors, aborting the database open process until the corruption is resolved.
+The error returned should be a distinct type that can easily be matched to 
+determine if the error is due to non-transient corruption or a transient error
+(such as network connectivity issues).
+
+#### Handling SST Corruption During Compaction
+Compaction is an asynchronous process that may encounter corrupted SST files
+during operation. If corruption is detected, the compaction process defers
+control to the `DBOptions.on_corruption` callback.
+
+- If the `DBOptions.on_corruption` callback returns an error, the compaction
+  operation aborts the current run.
+- If the user-defined `DBOptions.on_corruption` callback does not return an
+  error, compaction will ignore the corruption to the best of its ability,
+  which may result in data loss.
+
+Compaction will continue to attempt future runs, aborting each time corruption
+is encountered until the corruption is repaired or compaction is halted. If
+compaction were to never run again, it is possible that the operator might
+remain unaware that compaction has stopped and fail to restart it after
+repairing the corruption.
+
+#### Handling SST Corruption During Get
```

Comment:
can we add sections outlining how we want to handle corruption during compaction/gc? the use case is that the bit rot happens but is not detected by a `get` but rather by compaction/gc. Wondering if the callback still makes sense in that situation or if there's something else we should do

Reply:
I've added a section `#### Handling SST Corruption During Compaction`